### PR TITLE
WIP Make linear_regression_raw more sklearn compatible

### LINKED ISSUE #3787: FIX & ENH: linear_regression_raw solver sklearn style
The solver parameter in `mne.stats.linear_regression_raw` can now be set with a function. However, the `y` is transpose as compared to the expected equation `w = X^T Y`

Here is  the copy from the default solver used in `linear_regression_raw`:
```python
linalg.solve(a, X.T * y.T, sym_pos=True, overwrite_a=True, overwrite_b=True).T
```

For example if one use a sklearn's ridge, we would have to transpose y before passing the solver:

```python
from sklearn.linear_model.ridge import ridge_regression
solver = lambda X, Y: ridge_regression(X, Y.T, alpha=1.)
event_id = dict((c, ii) for ii, c in enumerate(categories))
evokeds = linear_regression_raw(raw, events, event_id, solver=solver)
```

I suggest a couple of things:
* transpose `Y` back to its original shape
* detect whether the solver has a 'fit' attribute, in which case it'd do `solver.fit(X, Y).coef_`
* add a dictionary to call the commonly used regressions (e.g. `solver='ridge'`)

### PR DESCRIPTION
**Pull request by jona-sassenhagen**
Sorry for being so slow @kingjr .

closes #3787

### CODE REVIEW

**Comment by kingjr** [comment]
I wouldn't change the default, you could keep cholevsky IMO

**Comment by jona-sassenhagen** [comment]
Oops yes, that was unintentional. Came from testing it. I will revert.

**Comment by jona-sassenhagen** [comment]
@Eric89GXL should I bother to make test cases to hit the remaining lines? They're:
- two error cases
- the Ridge solver

**Comment by jona-sassenhagen** [comment]
@kingjr can I get a +1?

**Comment by Eric89GXL** [comment]
Coverage is probably okay as is. We can improve it at the sprint.

File: mne/stats/regression.py [hunk 1]
```diff
@@ -244,19 +250,33 @@ def linear_regression_raw(raw, events, event_id=None, tmin=-.1, tmax=1,
...
     if isinstance(solver, string_types):
+        if solver not in ("cholesky", "ridge"):
+            raise ValueError("No such solver: {0}".format(solver))
+
+        if solver == 'ridge':
+            try:
+                from sklearn.linear_model.ridge import ridge_regression
+
+                def solver(X, y):
+                    return ridge_regression(X, y, solver='cholesky', alpha=1.)
```
Comment (agramfort) [inline mne/stats/regression.py]: you don't expose the alpha parameter? that seems to suggest alpha=1. is a good default which I doubt.

Comment (jona-sassenhagen) [inline mne/stats/regression.py]: It depends - this is not for prediction after all, you care about the model coefs, where you might prefer to have low bias. The main benefit of this is you can deal with colinear problems better.

Also the main point of this is to make user-supplanted solvers a bit easier.

I guess I can overload the `solver` kwarg to also allow a float, which would become the regularization parameter?

**Comment by agramfort** [comment]
I would not support the ridge string option but just update the docstring
to better explain
how to use a custom solver. It should even be in the narrative doc but I
will not force you to do it.

**Comment by jona-sassenhagen** [comment]
You mean, just a float option for ridge?

This PR is in response to @kingjr pointing out that the preferred usage seems to be using ridge anyways, asking for a str path to ridge.

Adding to the docs seems like a good task for the sprint :)

**Comment by agramfort** [comment]
then I would just expose an \alpha parameter that would default to 0 and
keep solver to cholesky etc.

we should not mix solver and estimator that as a solver parameter.

**Comment by jona-sassenhagen** [comment]
@kingjr what do you think?

**Comment by kingjr** [comment]
Mmmh, alex is right: we've confused solvers and estimators...

However, I'm -.5 on adding an alpha param; this would basically limit this function to the Ridge regression - as opposed to other approaches.

There's such a strong overlap between the Xdawn, linear_regression and chris' PR, that I have troubles seeing where we need to put the API efforts. I would recommend we wait until we all skype/sit together to decide how we structure ReceptiveField, linear_regression Xdawn etc. IMO, only of them should have a highly parameterizable API (i.e. specifying various estimators etc), whereas the other ones should be high-level conveniences.

tl;dr: wait

**Comment by jona-sassenhagen** [comment]
But in the meantime ..?

I think having Ridge with a default low alpha here as a str option is not a bad thing, and it's a tiny change. But regardless, the transpose of y should be done now.

**Comment by kingjr** [comment]
> the transpose of y should be done now.

+1

> I think having Ridge with a default low alpha here as a str option

WDYM?

**Comment by jona-sassenhagen** [comment]
I think in this function, a Ridge with alpha=1  - as the PR is right now - is not a bad default. It's not in a prediction context for now, given that that's probably gonna fall under the umbrella of Chris' function.

File: mne/stats/regression.py (hunk 1 above)
Comment (jaeilepp) [inline mne/stats/regression.py]: Maybe this should use ``try - except - else`` structure to avoid catching possible ``ImportError`` from ``ridge_regression``.

**Comment by jona-sassenhagen** [comment]
@kingjr where are we here? Should I just implement the transpose, dropping the str-to-ridge thing?

**Comment by kingjr** [comment]
IMO yes, just the transpose

File: mne/stats/regression.py [hunk 2]
```diff
@@ -227,8 +228,13 @@ def linear_regression_raw(raw, events, event_id=None, tmin=-.1, tmax=1,
...
         matrix X and the observation matrix Y, and returns the coefficient
-        matrix b; or a string. If str, must be ``'cholesky'``, in which case
-        the solver used is ``linalg.solve(dot(X.T, X), dot(X.T, y))``.
+        matrix b; or a string.
+        X is of shape (n_times, n_predictors * time_window_length).
+        y is of shape (n_channels, n_times).
+        If str, must be either ``'cholesky'``, in which
+        case the solver used is ``linalg.solve(dot(X.T, X), dot(X.T, y))``,
+        or ``'ridge'``, in which case importing `ridge_regression` from
+        sklearn is attempted; otherwise, defaults to the same solver as before.
```
Comment (wmvanvliet) [inline mne/stats/regression.py]: This is unclear to me. What do you mean by "defaults to the same solver as before." ?

File: mne/stats/regression.py (hunk 1 above)
Comment (wmvanvliet) [inline mne/stats/regression.py]: Why not use RidgeCV with `alpha=np.linspace(3, 3, 11)`? Finding the optimal `alpha` parameter is very cheap in scikit-learn's implementation of Ridge.

Comment (jona-sassenhagen) [inline mne/stats/regression.py]: RidgeCV can still be *very* slow for continuous-regression problems.

**Comment by Eric89GXL** [comment]
Okay to move this to 0.15?

**Comment by jona-sassenhagen** [comment]
Yeah sorry Eric ... It's a Code Sprint thing though. I must get it done at least for .15.

**Comment by kingjr** [comment]
+1 for after release; I have discussion point to address with Jona

File: mne/stats/regression.py [hunk 3]
```diff
@@ -266,7 +279,12 @@ def solver(X, y):
     X, data = _clean_rerp_input(X, data, reject, flat, decim, info, tstep)
 
     # solve linear system
-    coefs = solver(X, data)
+    coefs = solver(X, data.T)
+    if coefs.shape[0] != data.shape[0]:
+        msg = ("solver output has unexcepted shape. Supply a function that "
+               "returns coefficients in the form (n_targets, n_features), "
+               "where targets == channels.")
+        raise ValueError(msg)
```
Comment (Eric89GXL) [inline mne/stats/regression.py]: No need to separate these (and hurt coverage), can always do:
```
raise ValueError(
    ...)
```

**Comment by jona-sassenhagen** [comment]
Yeah @kingjr could you check quickly? This one is overlapping heavily with the other PR I'm working on right now, so I'd like to finish these one and begin dealing with the merge conflicts ASAP ...

File: mne/stats/regression.py [hunk 4]
```diff
@@ -244,19 +248,28 @@ def linear_regression_raw(raw, events, event_id=None, tmin=-.1, tmax=1,
            considerations. Psychophysiology, 52(2), 169-189.
     """
     if isinstance(solver, string_types):
+        if solver not in {"cholesky"}:
```
Comment (kingjr) [inline mne/stats/regression.py]: why dict?

File: mne/stats/regression.py [hunk 5]
```diff
@@ -244,19 +248,28 @@ def linear_regression_raw(raw, events, event_id=None, tmin=-.1, tmax=1,
...
         if solver == 'cholesky':
             def solver(X, y):
                 a = (X.T * X).toarray()  # dot product of sparse matrices
-                return linalg.solve(a, X.T * y.T, sym_pos=True,
+                return linalg.solve(a, X.T * y, sym_pos=True,
                                     overwrite_a=True, overwrite_b=True).T
-
-        else:
-            raise ValueError("No such solver: {0}".format(solver))
+    elif callable(solver):
```
Comment (kingjr) [inline mne/stats/regression.py]: is the possibility of using a callable an enh? or is this a fix?

File: mne/stats/regression.py [hunk 6]
```diff
@@ -244,19 +248,28 @@ def linear_regression_raw(raw, events, event_id=None, tmin=-.1, tmax=1,
...
+             "different format, it will give wrong results and might in "
+             "extreme cases crash your session.")
 
     # build data
     data, info, events = _prepare_rerp_data(raw, events, picks=picks,
                                             decim=decim)
 
+    if event_id is None:
+        event_id = dict((str(ii), v) for ii, v in
+                        enumerate(np.unique(events[:, 2])))
```
Comment (kingjr) [inline mne/stats/regression.py]: np.unique => set

File: mne/stats/regression.py [hunk 7]
```diff
@@ -266,7 +279,11 @@ def solver(X, y):
     X, data = _clean_rerp_input(X, data, reject, flat, decim, info, tstep)
 
     # solve linear system
-    coefs = solver(X, data)
+    coefs = solver(X, data.T)
+    if coefs.shape[0] != data.shape[0]:
+        raise ValueError("solver output has unexcepted shape. Supply a "
+                         "function that returns coefficients in the form "
+                         "(n_targets, n_features), where targets == channels.")
```
Comment (kingjr) [inline mne/stats/regression.py]: This error isn't tested is it?

File: mne/stats/tests/test_regression.py [hunk 8]
```diff
@@ -120,6 +121,8 @@ def test_continuous_regression_no_overlap():
                   events, event_id, tmin, tmax, decim=2)
 
 
+@requires_sklearn
+@testing.requires_testing_data
 def test_continuous_regression_with_overlap():
```
Comment (kingjr) [inline mne/stats/tests/test_regression.py]: add error testing: 
- solver string only in cholesky
- solver output has right shape
- non callable solver throw error

File: mne/stats/tests/test_regression.py [hunk 9]
```diff
@@ -19,6 +19,7 @@
 from mne.datasets import testing
 from mne.stats.regression import linear_regression, linear_regression_raw
 from mne.io import RawArray
+from mne.utils import requires_sklearn
```
Comment (kingjr) [inline mne/stats/tests/test_regression.py]: the imports should be relative for mne

File: mne/stats/regression.py (hunk 5 above)
Comment (jona-sassenhagen) [inline mne/stats/regression.py]: Well we warn cause the behavior is changed ...

File: mne/stats/regression.py (hunk 4 above)
Comment (jona-sassenhagen) [inline mne/stats/regression.py]: It's a set :D I was told I should use sets for membership tests.

File: mne/stats/tests/test_regression.py (hunk 9 above)
Comment (jona-sassenhagen) [inline mne/stats/tests/test_regression.py]: argh what did i do

Comment (Eric89GXL) [inline mne/stats/tests/test_regression.py]: this is fine it's in a test, not the module

File: mne/stats/regression.py (hunk 5 above)
Comment (jona-sassenhagen) [inline mne/stats/regression.py]: You could always supply a callable (this is why you raised the issue in the first place :)), I'm just adding the warning for the fix.

File: mne/stats/tests/test_regression.py [hunk 10]
```diff
@@ -135,3 +138,16 @@ def test_continuous_regression_with_overlap():
...
                     .data.flatten())
+
+    # test that sklearn solvers can be used
+    from sklearn.linear_model.ridge import ridge_regression
+    solver = lambda X, y: ridge_regression(X, y, alpha=1.)
+    assert_allclose(effect,
+                    linear_regression_raw(raw, events, tmin=0,
+                                          solver=solver)[1].data.flatten())
+
+	# test bad solvers
```
Comment (agramfort) [inline mne/stats/tests/test_regression.py]: indent pb :)

**Comment by agramfort** [comment]
besides LGTM

**Comment by agramfort** [comment]
CIs complain

**Comment by agramfort** [comment]
merged by rebase after fixing your tests
